### PR TITLE
2827 add filtering for courses to api

### PR DESCRIPTION
**Pull request by gpeng**
### Context

Find needs to be able to display a list of currently available courses and filter them by various filters.

### Changes proposed in this pull request

* Add scopes to `Course` to enable filtering
* Add a `CourseSearch` service that builds a `Course` scope based on the provided filter parameters.
* Add `/api/v3/recruitment_cycle/<year>/courses` endpoint
* Add indexes to `course_site` table for the enums on `SiteStatus`

### Guidance to review

Filter params are passed in the following format:

`/api/v3/recruitment_cycle/<year>/courses?filter[salary]=true&filter[qualifications]=pgce,pgde`

Out of scope in this PR

* Only returning 'published' courses
* Location filtering

### Checklist

- [x] Make sure all information from the Trello card is in here
- [x] Attach to Trello card
- [x] Rebased master
- [x] Cleaned commit history
- [ ] Tested by running locally

### CODE REVIEW

**Comment by gpeng** [comment]
[[S] Add filtering for courses to api part 1](https://trello.com/c/T2gCjkut/2827-s-add-filtering-for-courses-to-api-part-1)

File: app/controllers/api/v3/course_searches_controller.rb [hunk 1]
```diff
@@ -0,0 +1,10 @@
+module API
+  module V3
+    class CourseSearchesController < API::V3::ApplicationController
+      def index
+        course_scope = CourseSearch.call(filter: params[:filter], recruitment_cycle_year: params[:recruitment_cycle_year])
```
Comment (defong) [inline app/controllers/api/v3/course_searches_controller.rb]: may be use as default
`Settings.current_recruitment_cycle_year`

File: config/routes.rb [hunk 2]
```diff
@@ -91,9 +91,15 @@
...
+#                                        api_v3_recruitment_cycle_course GET    /api/v3/recruitment_cycles/:recruitment_cycle_year/courses/:id(.:format)                                                                 api/v3/course_searches#show
+#                                                                        PATCH  /api/v3/recruitment_cycles/:recruitment_cycle_year/courses/:id(.:format)                                                                 api/v3/course_searches#update
+#                                                                        PUT    /api/v3/recruitment_cycles/:recruitment_cycle_year/courses/:id(.:format)                                                                 api/v3/course_searches#update
+#                                                                        DELETE /api/v3/recruitment_cycles/:recruitment_cycle_year/courses/:id(.:format)                                                                 api/v3/course_searches#destroy
```
Comment (defong) [inline config/routes.rb]: apart from `GET|POST` I don't think there is need for the other verbs...

File: config/routes.rb [hunk 3]
```diff
@@ -91,9 +91,15 @@
 #                                                                        PUT    /api/v2/access_requests/:id(.:format)                                                                                                    api/v2/access_requests#update
 #                                                                        DELETE /api/v2/access_requests/:id(.:format)                                                                                                    api/v2/access_requests#destroy
 #                                                api_v2_build_new_course GET    /api/v2/build_new_course(.:format)                                                                                                       api/v2/courses#build_new
+#                                       api_v3_recruitment_cycle_courses GET    /api/v3/recruitment_cycles/:recruitment_cycle_year/courses(.:format)                                                                     api/v3/course_searches#index
```
Comment (defong) [inline config/routes.rb]: I think there we should expose `/course` and use that as the default for frontend

`recruitment_cycle_year` is nice but I don't think we have an actual need to expose it.

File: db/migrate/20200131104849_add_site_status_indexes.rb [hunk 4]
```diff
@@ -0,0 +1,9 @@
+class AddSiteStatusIndexes < ActiveRecord::Migration[6.0]
+  disable_ddl_transaction!
+
+  def change
+    add_index :course_site, :publish, algorithm: :concurrently
```
Comment (defong) [inline db/migrate/20200131104849_add_site_status_indexes.rb]: :+1: 

should annotate add to these bit as well?
``` ruby
# Indexes
#
#  IX_course_site_course_id  (course_id)
#  IX_course_site_site_id    (site_id)
#  xyz


class SiteStatus < ApplicationRecord
```

File: db/schema.rb [hunk 5]
```diff
@@ -119,7 +119,10 @@
     t.text "status"
     t.text "vac_status"
     t.index ["course_id"], name: "IX_course_site_course_id"
+    t.index ["publish"], name: "index_course_site_on_publish"
```
Comment (defong) [inline db/schema.rb]: before we forget 

just maybe

https://www.postgresql.org/docs/9.6/indexes-multicolumn.html

```
  scope :findable, -> { status_running.published_on_ucas }


```
@timabell fyi

File: app/services/course_search.rb [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+  end
+
+  class << self
+    def call(**args)
+      new(args).call
+    end
+  end
+
+  def call
+    scope = Course
```
Comment (defong) [inline app/services/course_search.rb]: Just wondering does scope need to be ordered to be more performant?

ie

```ruby
faster = course
      .with_recruitment_cycle(recruitment_cycle_year)
      .findable

# vs 

slower = course
      .findable
      .with_recruitment_cycle(recruitment_cycle_year)

```

File: config/routes.rb [hunk 7]
```diff
@@ -175,6 +181,7 @@
       resources :recruitment_cycles,
                 only: :show,
                 param: :year do
+        resources :courses, controller: "course_searches"
```
Comment (misaka) [inline config/routes.rb]: For providers, the search was incorporated as part of the [existing `CourseController`](https://github.com/DFE-Digital/teacher-training-api/pull/1144/files#diff-63584eeba1d49b4829082079274d0447R9), I like this because we should support searches, etc, on any `Course` index.

Can we just add a `courses` resource here that uses the existing `CourseController` and add search support to it?

File: app/services/course_search.rb (hunk 6 above)
Comment (misaka) [reply app/services/course_search.rb]: No, these should be functionally equivalent.

Don't forget these are `ActiveRecord::Relation` objects, which means that a) the accumulate the parameters and only execute them once the results are requested, and b) we can examine what the SQL looks like:

```sql
> Course.findable.with_recruitment_cycle('2020').to_sql
SELECT "course".* FROM "course"
       INNER JOIN "course_site" ON "course_site"."course_id" = "course"."id"
       INNER JOIN "provider" ON "provider"."id" = "course"."provider_id"
       INNER JOIN "recruitment_cycle" ON "recruitment_cycle"."id" = "provider"."recruitment_cycle_id"
       WHERE "course_site"."status" = 'R'
             AND "course_site"."publish" = 'Y'
             AND "recruitment_cycle"."year" = '2020'


> Course.with_recruitment_cycle('2020').findable.to_sql
SELECT "course".* FROM "course"
       INNER JOIN "provider" ON "provider"."id" = "course"."provider_id"
       INNER JOIN "recruitment_cycle" ON "recruitment_cycle"."id" = "provider"."recruitment_cycle_id"
       INNER JOIN "course_site" ON "course_site"."course_id" = "course"."id"
       WHERE "recruitment_cycle"."year" = '2020'
             AND "course_site"."status" = 'R'
             AND "course_site"."publish" = 'Y'

```

File: config/routes.rb (hunk 3 above)
Comment (misaka) [reply config/routes.rb]: Ya, I think that that makes sense too. `find` doesn't need to know what the current cycle is.

File: spec/models/course_spec.rb [hunk 8]
```diff
@@ -384,6 +384,230 @@
...
+        expect(subject).to contain_exactly(course_school_direct_salaried_training_programme)
+      end
+    end
+
+    describe ".with_qualifications" do
+      let(:course_qts) { create(:course, :resulting_in_qts) }
+      let(:course_pgce_with_qts) { create(:course, :resulting_in_pgce_with_qts) }
+      let(:course_pgde_with_qts) { create(:course, :resulting_in_pgde_with_qts) }
+      let(:course_pgce) { create(:course, :resulting_in_pgce) }
+      let(:course_pgde) { create(:course, :resulting_in_pgde) }
```
Comment (misaka) [inline spec/models/course_spec.rb]: What do you think about adding these, and others, to `TestDataCache`? It'd be nice to start to populate that with commonly-used course types, instead of having to back-fill more and more of these.

File: db/schema.rb (hunk 5 above)
Comment (gpeng) [reply db/schema.rb]: Discussed offline. The individual indices will be fine here with this table size but this is a potential optimisation in the future.

File: config/routes.rb (hunk 7 above)
Comment (gpeng) [reply config/routes.rb]: I've updated this after offline discussion. This is now `/api/V3/courses` and only routes `index`.

**Comment by gpeng** [comment]
We discussed adding search to the existing courses controller offline and have decided not to do this because:

* We don't currently have a user need for search on that endpoint
* it would require adding provider filtering to the `CourseSearch` class which is additional complexity we don't need yet
* The `CourseSearchesController` class will be easy to incorporate into `CoursesController` in the future if we need to without breaking the API. (It would also be difficult to remove search from `CoursesController` if we added it and realised we didn't have a user need for it after all.

File: spec/models/course_spec.rb (hunk 8 above)
Comment (gpeng) [reply spec/models/course_spec.rb]: I've added the `TestDataCache` as discussed offline. Turns out, there were only a small number that worked with it as most in this PR require parameters. I've left it in for now but I'm not sure the overhead is worth it in the unit tests as in this PR. I wonder if we'd be best benchmarking the specs to identify the slowest one and then trying `TestDataCache` on that and getting some data on the performance and seeing the development overhead to add it. We could also try some other approaches on the same file and compare before we spread it too far through the app.

File: app/services/course_search.rb [hunk 9]
```diff
@@ -0,0 +1,50 @@
+class CourseSearch
```
Comment (misaka) [inline app/services/course_search.rb]: Our other services use the suffix `Service`, can we bring this one in line? Can be in another PR.
